### PR TITLE
JENKINS-25048: Added filters check for notifyCommit event

### PR DESCRIPTION
**Pull request by pbaranchikov**
Continuation of #263
I do not really know, whether getProject(Item item) method should be used, or simple cast to AbstractProject should be used in form
<code>
if (Item instanceof AbstractProject) {
   AbstractProject <?,?> projectObject = (AbstractProject<?,?>) item;
}
</code>

### CODE REVIEW

File: src/main/java/hudson/plugins/git/GitSCM.java [hunk 1]
```diff
@@ -1552,6 +1552,57 @@ protected FilePath workingDirectory(Job<?,?> context, FilePath workspace, EnvVar
...
+        }
+        final FilePath workspace = lastBuild.getWorkspace();
+        final EnvVars environment = GitUtils.getPollEnvironment(project, workspace, null, listener);
+
+        final FilePath workingDirectory = workingDirectory(project, workspace, environment,
+                listener);
+        // which node is this workspace from?
+        // there should be always one match, but just in case we initialize n to
+        // a non-null value
+        Node n = Jenkins.getInstance();
```
Comment (KostyaSha) [inline src/main/java/hudson/plugins/git/GitSCM.java]: usually variable named "jenkins" or "j" or "instance"

Comment (KostyaSha) [inline src/main/java/hudson/plugins/git/GitSCM.java]: Hmm... looks like a weird trick..

**Comment by dspartz** [comment]
When will this PR be merged and new version of git plugin built?

I've cherry picked the #263 PR, and applied to 2.2.10 months ago. We want to move to 2.4.0, but after fixing the patch for that version, the added tests for the PR fail:

Tests run: 58, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 224.999 sec <<< FAILURE! - in hudson.plugins.git.GitSCMTest
testNotifyCommitMixedWithPoll(hudson.plugins.git.GitSCMTest)  Time elapsed: 1.163 sec  <<< FAILURE!
junit.framework.AssertionFailedError: notifyCommit should detect commit 1
	at junit.framework.Assert.fail(Assert.java:57)
	at junit.framework.Assert.assertTrue(Assert.java:22)
	at junit.framework.TestCase.assertTrue(TestCase.java:192)
	at hudson.plugins.git.GitSCMTest.testNotifyCommitMixedWithPoll(GitSCMTest.java:1890)

testNotifyCommitOnExcluded(hudson.plugins.git.GitSCMTest)  Time elapsed: 1.073 sec  <<< FAILURE!
junit.framework.AssertionFailedError: notifyCommit should detect commit 2
	at junit.framework.Assert.fail(Assert.java:57)
	at junit.framework.Assert.assertTrue(Assert.java:22)
	at junit.framework.TestCase.assertTrue(TestCase.java:192)
	at hudson.plugins.git.GitSCMTest.testNotifyCommitOnExcluded(GitSCMTest.java:1873)


I'd attach the patch I used, but don't know how to :-( (I manually found the correct line numbers to insert/delete/update to fit the PR in)

**Comment by MarkEWaite** [comment]
It will likely be a month or more before this is merged into the plugin.  Each pull request commonly takes a week or more to evaluate, intermixed with handling user questions and bug reports.  This particular area of the code has been challenging, so this one will likely take longer to evaluate than typical.

If you need it sooner, you could help the project as a whole by pulling these changes onto a branch based on git plugin 2.4.0 and then use it actively.  Report your results so that the plugin developers gain the benefit of active users reporting their experiences before the change is made visible to the larger community of users.

**Comment by jglick** [comment]
Patch seems dubious. Only works for `AbstractProject` (i.e., no workflow), and only for a project whose last build happens to still have a workspace (i.e., no cloud slaves).

IIUC `/git/notifyCommit` is just supposed to _schedule polling_, not necessarily start a build, so any checks should be done during polling not now.

**Comment by carpnick** [comment]
@jglick Agree.  Asking for trouble doing it here.  Maybe generate a polling event that triggers the other more generic code.

**Comment by cniweb** [comment]
Is there a solution for this?

**Comment by MarkEWaite** [comment]
@cniweb no, I'm not aware of any  solution.

**Pull request closed without merging**